### PR TITLE
F/injective-math improvements

### PR DESCRIPTION
**Pull request by DrunkenRandomWalker**
1. replaced `bigint` with `primitive-types`, because `bigint` is deprecated.
2. improved `exp`, `pow`, and `ln` numerical precision to double precision.
3. added newton's method, discrete_newton's method, halley's method for root finding.

### CODE REVIEW

File: packages/injective-math/Cargo.toml [hunk 1]
```diff
@@ -6,7 +6,7 @@ license     = "Apache-2.0"
 name        = "injective-math"
 readme      = "README.md"
 repository  = "https://github.com/InjectiveLabs/cw-injective/tree/master/packages/injective-math"
-version     = "0.1.23"
+version     = "0.1.24"
```
Comment (maxrobot) [inline packages/injective-math/Cargo.toml]: I think this a major release especially with the bigint change

File: packages/injective-math/src/fp_decimal/arithmetic.rs [hunk 2]
```diff
@@ -77,30 +81,31 @@ impl FPDecimal {
...
-        let num = FPDecimal::ONE.num.full_mul(x.num) / y.num.into();
+        let num = FPDecimal::ONE.num.full_mul(x.num) / U512::try_from(y.num).unwrap();
         if num.is_zero() {
-            return FPDecimal::zero();
+            return FPDecimal::ZERO;
         }
 
         FPDecimal {
-            num: num.into(),
+            num: U256::try_from(num).unwrap(),
```
Comment (maxrobot) [inline packages/injective-math/src/fp_decimal/arithmetic.rs]: can't this panic using unrwap?

File: packages/injective-math/src/fp_decimal/exp.rs [hunk 3]
```diff
@@ -66,21 +759,60 @@ impl FPDecimal {
...
+    //     for i in 1..((2 * FPDecimal::DIGITS + 1) as u64) {
+    //         d = (d * x) / (FPDecimal::ONE.num * U256([i, 0, 0, 0]));
+    //         tr = tr + d;
+    //     }
+    //     let val = FPDecimal::_mul(FPDecimal { num: tr, sign: 1 }, r);
+    //     if a.sign == 0 {
+    //         return FPDecimal::reciprocal(val);
+    //     }
+    //     val
+    // }
```
Comment (maxrobot) [inline packages/injective-math/src/fp_decimal/exp.rs]: why leaving this commented?

File: packages/injective-math/src/fp_decimal/exp.rs [hunk 4]
```diff
@@ -1092,53 +1824,215 @@ impl FPDecimal {
...
+    //             Err(_) => panic!("Multiplication overflow"),
+    //         }
+    //     } else {
+    //         match self.checked_negative_pow(exponent) {
+    //             Ok(value) => value,
+    //             Err(_) => panic!("Multiplication overflow"),
+    //         }
+    //     }
+    // }
+
```
Comment (maxrobot) [inline packages/injective-math/src/fp_decimal/exp.rs]: so is there no power function anymore?

File: packages/injective-math/src/lib.rs [hunk 5]
```diff
@@ -1,9 +1,11 @@
 pub mod fp_decimal;
+pub mod root_findings;
```
Comment (maxrobot) [inline packages/injective-math/src/lib.rs:2]: fix these errors

**Review by maxrobot — CHANGES_REQUESTED** [review]
Seems like a good step to making the package better but can we:

1. make sure there is a consistent interface

Meaning that all public functions are not underscored, and if they are not to be public the to perform that at the function definition level

Keep mathematical operations like power, exp, sqrt available.

2. Also remove uncommented code

3. There is use of unwrap which I think is unsafe, please make sure we are handling unwraps appropriately.

File: packages/injective-math/src/fp_decimal/exp.rs (hunk 4 above)
Comment (DrunkenRandomWalker) [reply packages/injective-math/src/fp_decimal/exp.rs]: we have `pow` function its in line 58

File: packages/injective-math/src/lib.rs (hunk 5 above)
Comment (DrunkenRandomWalker) [reply packages/injective-math/src/lib.rs:2]: i just pushed root finding algos

File: packages/injective-math/src/fp_decimal/exp.rs (hunk 3 above)
Comment (DrunkenRandomWalker) [reply packages/injective-math/src/fp_decimal/exp.rs]: fixed

File: packages/injective-math/src/fp_decimal/arithmetic.rs (hunk 2 above)
Comment (DrunkenRandomWalker) [reply packages/injective-math/src/fp_decimal/arithmetic.rs]: yes, it can panic.

In `Bigint` used `U512.into()`.
https://github.com/paritytech/bigint/blob/7e71521a61b009afc94c91135353102658550d42/src/uint.rs#L1044C1-L1058C2

It can panic too. So I decided to be consistent with `Bigint`

File: packages/injective-math/Cargo.toml (hunk 1 above)
Comment (DrunkenRandomWalker) [reply packages/injective-math/Cargo.toml]: I bumped version to 0.2.0

**Comment by DrunkenRandomWalker** [comment]
> Seems like a good step to making the package better but can we:
> 
>     1. make sure there is a consistent interface
> 
> 
> Meaning that all public functions are not underscored, and if they are not to be public the to perform that at the function definition level
> 
> Keep mathematical operations like power, exp, sqrt available.
> 
>     2. Also remove uncommented code
> 
>     3. There is use of unwrap which I think is unsafe, please make sure we are handling unwraps appropriately.

1 and 2 are done.

File: packages/injective-math/src/fp_decimal/log.rs [hunk 6]
```diff
@@ -517,19 +686,79 @@ impl FPDecimal {
         a.ln() / base.ln()
     }
 
-    /// natural logarithm
+    pub fn log(&self, base: FPDecimal) -> FPDecimal {
+        assert!(base > FPDecimal::ZERO);
+        if *self == FPDecimal::ONE {
+            return FPDecimal::ZERO;
+        }
+        if *self == FPDecimal::ZERO {
```
Comment (gorgos) [inline packages/injective-math/src/fp_decimal/log.rs]: you can use`is_zero()` here

File: packages/injective-math/src/fp_decimal/log.rs [hunk 7]
```diff
@@ -517,19 +686,79 @@ impl FPDecimal {
...
+            // FIXME should be an undefined, not sure if it will be better to just add assert!(b>0)
```
Comment (gorgos) [inline packages/injective-math/src/fp_decimal/log.rs]: I recommend returning `StdResult<FPDecimal>`

File: packages/injective-math/src/fp_decimal/arithmetic.rs [hunk 8]
```diff
@@ -18,19 +20,21 @@ impl FPDecimal {
                 sign: x.sign,
             };
         }
-        let mut sign = y.sign;
         if y.num == x.num {
-            sign = 1;
+            return FPDecimal::ZERO;
```
Comment (gorgos) [inline packages/injective-math/src/fp_decimal/arithmetic.rs:24]: how does that make any sense?

File: packages/injective-math/src/fp_decimal/arithmetic.rs (hunk 2 above)
Comment (gorgos) [reply packages/injective-math/src/fp_decimal/arithmetic.rs]: well in the case of division it cannot happen right? except for the one edge case of: `MAX_FPDECIMAL / -1`

**Review by gorgos — CHANGES_REQUESTED** [review]
looks mostly good, although I cant go through the math in detail of course, hopefully the tests and us using the new version will be sufficient

File: packages/injective-math/src/fp_decimal/arithmetic.rs (hunk 8 above)
Comment (DrunkenRandomWalker) [reply packages/injective-math/src/fp_decimal/arithmetic.rs:24]: The x and y have opposite sign in here.

File: packages/injective-math/src/fp_decimal/arithmetic.rs (hunk 2 above)
Comment (DrunkenRandomWalker) [reply packages/injective-math/src/fp_decimal/arithmetic.rs]: Yes, I think overflow can only happen in `MIN_FPDecimal/-1`. because `abs(Min) = Max +1`

File: packages/injective-math/src/fp_decimal/arithmetic.rs (hunk 8 above)
Comment (gorgos) [reply packages/injective-math/src/fp_decimal/arithmetic.rs:24]: So why not move the check higher above `if x.num > y.num {` ?

Comment (DrunkenRandomWalker) [reply packages/injective-math/src/fp_decimal/arithmetic.rs:24]: I thought this will save some cpu cycles

**Review by gorgos — APPROVED** [review]
LGTM

**Review by maxrobot — APPROVED** [review]
